### DISCUSSION TITLE
Save playlist/track and restart after alarm

### DISCUSSION THREAD

**Discussion by 4remu** [hklages/node-red-contrib-sonos-plus#226] [Q&A]
Hello

I want to use the alarm in the sonos app, to wake up my daughter in the morning. She is listening audiobooks on this box. Sadly, after the sonos alarm her playlist starts everytime from beginning. Is there an opportunity to solve this problem with this node? Thanks for help

**Comment by hklages** [comment]
Hi - thats a very nice use case. Lets try to make it happen. 

How did you set up the alarm? What music do you use - another playlist, a radio station, chime?
Regarding the audiobook: Is it just a simple playlist stored in the SONOS Queue?
What happens after the alarm is played. Is there a break? When should the audiobook be again available?

Idea - just to have a first draft for discussion. 
We can save the queue (means the playlist of all items of the audiobook) and with another node (node-red-contrib-sonos-events) we can store the last played song of the playlist in a Node-RED flow variable. Alternatively you can check every 20 second what queue item is playing - maybe together with the position in that part of  the audio book. After the alarm finishes we can then restore the queue and position quitely and when the Sonos play button is pushed again, the audiobook continues with the stored data.

**Reply by 4remu** [reply]
Thanks for your reply. My daughter put some Audiobook as an album on her waiting list. This list is around 150 titles long, and is not a playlist. just a SONOS queue. She is setting up the alarm  in the SONOS app under settings-->Speaker--> Alarm. The Alarm is just the standad SONOS Ringtone. First i tried with a playlist, but there it deleted her own queue which she is listening. With choosing the standard ringtone, her queue will also exists after stoping the alarm. But then her queue starts at beginning. 

What do you think, does your idea just work with a playlist?

**Comment by hklages** [comment]
Hi. 
Here is a flow to capture the last position in queue (every 20 seconds). The result is stored in a flow variable. 
You have to make flow variables persistant (otherwise after reboot they are lost). 
After your alarm had fired you can set the position in queue and play it with group.play.track. 
If that is not "exact"enoght you can do a group.create.snap and after the alarm a group.restore.snap together with group.play.queue. 

It will take a few hours to design a reliable flow - but should be possible. 

```
[{"id":"ace887da2d93673d","type":"inject","z":"94388f7bde18c5f3","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"20","crontab":"","once":true,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":170,"y":140,"wires":[["6b0c4ba1a428a70f"]]},{"id":"74565157dbcaa9dd","type":"sonos-universal","z":"94388f7bde18c5f3","confignode":"5a07e409.17a1ec","command":"group.get.trackplus","state":"","stateType":"str","avoidCheckPlayerAvailability":false,"name":"","x":893.4285714285714,"y":140,"wires":[["96401e9f110eb849"]]},{"id":"4bc425ac967f9708","type":"debug","z":"94388f7bde18c5f3","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","statusVal":"","statusType":"auto","x":1451,"y":140,"wires":[]},{"id":"238d0c4aca738f35","type":"rbe","z":"94388f7bde18c5f3","name":"","func":"rbe","gap":"","start":"","inout":"out","septopics":true,"property":"payload","topi":"topic","x":1285.142857142857,"y":140,"wires":[["4bc425ac967f9708","7fb42a611136a77d"]]},{"id":"96401e9f110eb849","type":"change","z":"94388f7bde18c5f3","name":"","rules":[{"t":"set","p":"payload","pt":"msg","to":"payload.positionData.Track","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":1109.2857142857142,"y":140,"wires":[["238d0c4aca738f35"]]},{"id":"7fb42a611136a77d","type":"change","z":"94388f7bde18c5f3","name":"","rules":[{"t":"set","p":"lastPositionInQueue","pt":"flow","to":"payload","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":1501,"y":180,"wires":[[]]},{"id":"6b0c4ba1a428a70f","type":"sonos-universal","z":"94388f7bde18c5f3","confignode":"5a07e409.17a1ec","command":"group.get.state","state":"","stateType":"str","avoidCheckPlayerAvailability":false,"name":"","x":365.85714285714283,"y":140,"wires":[["c4bd928db7782986"]]},{"id":"c4bd928db7782986","type":"switch","z":"94388f7bde18c5f3","name":"playing","property":"payload.playbackstate","propertyType":"msg","rules":[{"t":"eq","v":"playing","vt":"str"}],"checkall":"true","repair":false,"outputs":1,"x":551.7142857142857,"y":140,"wires":[["be5632ffaf405fda"]]},{"id":"be5632ffaf405fda","type":"switch","z":"94388f7bde18c5f3","name":"","property":"payload.queueActivated","propertyType":"msg","rules":[{"t":"true"}],"checkall":"true","repair":false,"outputs":1,"x":707.5714285714286,"y":140,"wires":[["74565157dbcaa9dd"]]},{"id":"5a07e409.17a1ec","type":"sonos-config","name":"yourPlayer","serialnum":"","ipaddress":"192.168.178.51"}]
```

**Comment by bednim** [comment]
has anyone got it working by any chance? this would be really great functionality to have. I am getting all the info with important RelTime but how would one inject it back into player?
Thank you
